Sync/BatchingClient.swift

```diff
@@ -0,0 +1,230 @@
+ *  by the info/configuration endpoint.
+ */
+private struct Payload {
+    let config: InfoConfiguration
+    private(set) var records: [UploadRecord] = []
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+
+    init(config: InfoConfiguration) {
+        self.config = config
+    }
+
+    /** Returns true if the record was added to this payload; false otherwise. */
+    mutating func add(record: UploadRecord) -> Bool {
+        if (record.size + bytes + 1) > config.maxPostBytes {
+            return false
+        }
+
+        if count >= config.maxPostRecords {
+            return false
+        }
+
+        records.append(record)
+        count += 1
+        bytes += record.size + 1     // For newline.
+        return true
+    }
+}
+
+/**
+ * Used to keep track of number of bytes and records sent as part of a batch
+ */
+private struct BatchMeta {
+    let config: InfoConfiguration
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+
+    init(config: InfoConfiguration) {
+        self.config = config
+    }
+
+    mutating func add(payload: Payload) -> Bool {
+        guard (bytes + payload.bytes <= config.maxTotalBytes) &&
+              (count + payload.count < config.maxTotalRecords) else
+        {
+            return false
+        }
+
+        count += payload.count
+        bytes += payload.bytes
+        return true
+    }
+}
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private(set) var ifUnmodifiedSince: Timestamp?
+
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    private var batchMeta: BatchMeta
+    private var batchToken: BatchToken?
+    private var currentPayload: Payload
+    private var onCollectionUploaded: (POSTResult -> DeferredTimestamp)
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult -> DeferredTimestamp))
+    {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.batchMeta = BatchMeta(config: config)
+        self.currentPayload = Payload(config: config)
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        return walk(records, f: addRecord)
```

Comment:
I like how this breaks down the operations logically into something that more resembles your conveyor belt analogy you used before. Pull off some records, bunch them up, add them while respecting the state of uploading. This would also clean up the current `addRecord` call but separating the operations a bit. I'll play around with this approach and see what I can do along with the other comments in the PR about atomically swapping out the payload addition.